File: C++/hash.cpp
#include <iostream>
#include <string>
#include <chrono>

using namespace std;
using namespace chrono;

constexpr int ITERATIONS = 1000000;
constexpr size_t STR_LEN = 1000;

unsigned int hash_sdbm(const char* str, size_t len) {
    unsigned int hash = 0;
    for (size_t i = 0; i < len; i++) {
        hash = str[i] + (hash << 6) + (hash << 16) - hash;
    }
    return hash;
}

unsigned int hash_djb2(const char* str, size_t len) {
    unsigned int hash = 5381;
    for (size_t i = 0; i < len; i++) {
        hash = ((hash << 5) + hash) + str[i];
    }
    return hash;
}

unsigned int hash_fnv(const char* str, size_t len) {
    unsigned int hash = 2166136261u;
    for (size_t i = 0; i < len; i++) {
        hash ^= str[i];
        hash *= 16777619u;
    }
    return hash;
}

int main() {
    cout << "C++ Hashing Benchmark (" << ITERATIONS << " iterations)\n";
    
    string data(STR_LEN, 'a');
    volatile unsigned int result = 0;
    
    auto t0 = high_resolution_clock::now();
    for (int i = 0; i < ITERATIONS; i++) {
        result += hash_sdbm(data.c_str(), STR_LEN);
    }
    auto t1 = high_resolution_clock::now();
    double ms = duration<double, milli>(t1 - t0).count();
    cout << "SDBM: " << ms << " ms (" << (ITERATIONS / (ms / 1000.0)) << " ops/sec)\n";
    
    t0 = high_resolution_clock::now();
    for (int i = 0; i < ITERATIONS; i++) {
        result += hash_djb2(data.c_str(), STR_LEN);
    }
    t1 = high_resolution_clock::now();
    ms = duration<double, milli>(t1 - t0).count();
    cout << "DJB2: " << ms << " ms (" << (ITERATIONS / (ms / 1000.0)) << " ops/sec)\n";
    
    t0 = high_resolution_clock::now();
    for (int i = 0; i < ITERATIONS; i++) {
        result += hash_fnv(data.c_str(), STR_LEN);
    }
    t1 = high_resolution_clock::now();
    ms = duration<double, milli>(t1 - t0).count();
    cout << "FNV: " << ms << " ms (" << (ITERATIONS / (ms / 1000.0)) << " ops/sec)\n";
    
    if (result == 0) cout << "";
    return 0;
}

File: Rust/hash.rs
use std::time::Instant;

const ITERATIONS: usize = 1_000_000;
const STR_LEN: usize = 1000;

fn hash_sdbm(s: &[u8]) -> u32 {
    let mut hash: u32 = 0;
    for &c in s {
        hash = c as u32 + (hash << 6) + (hash << 16) - hash;
    }
    hash
}

fn hash_djb2(s: &[u8]) -> u32 {
    let mut hash: u32 = 5381;
    for &c in s {
        hash = ((hash << 5) + hash) + c as u32;
    }
    hash
}

fn hash_fnv(s: &[u8]) -> u32 {
    let mut hash: u32 = 2166136261;
    for &c in s {
        hash ^= c as u32;
        hash = hash.wrapping_mul(16777619);
    }
    hash
}

fn main() {
    println!("Rust Hashing Benchmark ({} iterations)", ITERATIONS);
    
    let data: Vec<u8> = vec![b'a'; STR_LEN];
    let mut result: u32 = 0;
    
    let t0 = Instant::now();
    for _ in 0..ITERATIONS {
        result += hash_sdbm(&data);
    }
    let t1 = Instant::now();
    let ms = (t1 - t0).as_secs_f64() * 1000.0;
    println!("SDBM: {:.2} ms ({:.0} ops/sec)", 
        ms, ITERATIONS as f64 / (ms / 1000.0));
    
    let t0 = Instant::now();
    for _ in 0..ITERATIONS {
        result += hash_djb2(&data);
    }
    let t1 = Instant::now();
    let ms = (t1 - t0).as_secs_f64() * 1000.0;
    println!("DJB2: {:.2} ms ({:.0} ops/sec)", 
        ms, ITERATIONS as f64 / (ms / 1000.0));
    
    let t0 = Instant::now();
    for _ in 0..ITERATIONS {
        result += hash_fnv(&data);
    }
    let t1 = Instant::now();
    let ms = (t1 - t0).as_secs_f64() * 1000.0;
    println!("FNV: {:.2} ms ({:.0} ops/sec)", 
        ms, ITERATIONS as f64 / (ms / 1000.0));
    
    if result == 0 { println!(""); }
}
